### PR TITLE
Handle new 1P/3P schema

### PR DESCRIPTION
**Pull request by hestolz**
Handle new schema format, while retaining back-compat (disallow mix-and-match). Example of new schema below.

`settings`:
```
{ 
  "genevaConfiguration": { 
    "enable": true 
  }, 
  "azureMonitorConfiguration": { 
    "enable": true, 
    "configuration": { 
      "authentication": { 
        "managedIdentity": { 
          "identifier-name": "value", 
          "identifier-value": "value" 
        } 
      }, 
      "proxy": { 
        "mode": "application", 
        "address": "http://[address]:[port]", 
        "auth": false 
      } 
    } 
  } 
} 
```

`protectedSettings`:
```
{ 
  "azureMonitorConfiguration": { 
    "proxy": { 
      "username": "[username]", 
      "password": "[password]" 
    } 
  } 
} 
```

### CODE REVIEW

File: AzureMonitorAgent/agent.py [hunk 1]
```diff
@@ -388,16 +388,22 @@ def enable():
...
-    global AMAServiceStartCommand, AMAServiceRestartCommand, AMAServiceStatusCommand
 
-    # start the metrics process if its not already running
     public_settings, protected_settings = get_settings()
 
-    if (protected_settings is None or len(protected_settings) == 0) or (public_settings is not None and "proxy" in public_settings and "mode" in public_settings.get("proxy") and public_settings.get("proxy").get("mode") == "application"):
+    # start the metrics watcher if its not already running
+    if ((protected_settings is None or len(protected_settings) == 0) or
+        (public_settings is not None and "proxy" in public_settings and "mode" in public_settings.get("proxy") and public_settings.get("proxy").get("mode") == "application") or
+        (public_settings is not None and public_settings.get("azureMonitorConfiguration") is not None and public_settings.get("azureMonitorConfiguration").get("ensure") == True)):
```
Comment (NarineM) [inline AzureMonitorAgent/agent.py:397]: Do you want to move "azureMomitorConfiguration" to be a const? I see it repeated , same for ""azuremonitoragent"

Comment (NarineM) [reply AzureMonitorAgent/agent.py:397]: Looks good. I can sign off assuming upgrade with old settings and brand new install, both the paths were tested?

**Review by simathih — DISMISSED** [review]

File: AzureMonitorAgent/agent.py (hunk 1 above)
Comment (simathih) [reply AzureMonitorAgent/agent.py:397]: +1

**Review by simathih — APPROVED** [review]

File: AzureMonitorAgent/agent.py (hunk 1 above)
Comment (hestolz) [reply AzureMonitorAgent/agent.py]: I tested it as follows
- next-gen geneva AND azure monitor config
- next-gen geneva config
- next-gen azure monitor config
- next-gen azure monitor config with parameters (e.g. managed identity)
- legacy geneva config
- legacy autoconfig
- legacy azure monitor config
- mixed legacy and next-gen (results in error as expected)

File: AzureMonitorAgent/agent.py [hunk 2]
```diff
@@ -588,54 +484,197 @@ def enable():
...
+
+        if MONITORING_ROLE != "":
+            default_configs["MONITORING_ROLE"] = MONITORING_ROLE
+
+        if MONITORING_TENANT != "":
+            default_configs["MONITORING_ROLE_INSTANCE"] = MONITORING_ROLE_INSTANCE
+
+def handle_mcs_config(public_settings, protected_settings, default_configs):
+    """
+    Populate the defualts for MCS mode
```
Comment (AaronRM) [inline AzureMonitorAgent/agent.py:637]: >defualts [](http://example.com/codeflow?start=17&length=8)

Nit: spelling

**Review by AaronRM — APPROVED** [review]
:shipit:

**Review by AaronRM — APPROVED** [review]

File: AzureMonitorAgent/agent.py [hunk 3]
```diff
@@ -588,54 +484,197 @@ def enable():
...
+
+        if "auth" in public_settings.get("proxy") and public_settings.get("proxy").get("auth") == "true":
+            if protected_settings is not None and "proxy" in protected_settings and "username" in protected_settings.get("proxy") and "password" in protected_settings.get("proxy"):
+                default_configs["MDSD_PROXY_USERNAME"] = protected_settings.get("proxy").get("username")
+                default_configs["MDSD_PROXY_PASSWORD"] = protected_settings.get("proxy").get("password")
+            else:
+                log_and_exit("Enable", MissingorInvalidParameterErrorCode, 'Parameter "username" and "password" not in proxy protected setting')
+
+    # add managed identity settings if they were provided
+    identifier_name, identifier_value, error_msg = get_managed_identity()
```
Comment (canfikret) [inline AzureMonitorAgent/agent.py:661]: Nit: use managed_id_name, managed_id_value for better context compared to identifier_name, identifier_value

File: AzureMonitorAgent/agent.py [hunk 4]
```diff
@@ -113,21 +113,17 @@ def __str__(self):
...
-AMAUninstallCommand = ''
-AMAServiceStartCommand = ''
-AMAServiceRestartCommand = ''
-AMAServiceStopCommand = ''
-AMAServiceStatusCommand = ''
-
 # Error codes
 GenericErrorCode = 1
 UnsupportedOperatingSystem = 51
 IndeterminateOperatingSystem = 51
```
[the remark this answers is not in the document]
Comment (hestolz) [reply AzureMonitorAgent/agent.py:119]: how come? https://github.com/Azure/azure-marketplace/wiki/Extension-Build-Notes-Best-Practices#error-codes-and-messages-output-to-stderr

**Pull request merged**